frame/support/test/tests/pallet_compatibility_instance.rs

```diff
@@ -0,0 +1,311 @@
+			type Error = Error<T, I>;
+			fn deposit_event() = default;
+			const SomeConst: T::Balance = T::SomeConst::get();
+
+			#[weight = <T::Balance as Into<Weight>>::into(new_value.clone())]
+			fn set_dummy(origin, #[compact] new_value: T::Balance) {
+				ensure_root(origin)?;
+
+				<Dummy<T, I>>::put(&new_value);
+				Self::deposit_event(RawEvent::Dummy(new_value));
+			}
+
+			fn on_initialize(_n: T::BlockNumber) -> Weight {
+				<Dummy<T, I>>::put(T::Balance::from(10));
+				10
+			}
+
+			fn on_finalize(_n: T::BlockNumber) {
+				<Dummy<T, I>>::put(T::Balance::from(11));
+			}
+		}
+	}
+
+	decl_error! {
+		pub enum Error for Module<T: Trait<I>, I: Instance> {
+			/// Some wrong behavior
+			Wrong,
+		}
+	}
+}
+
+#[frame_support::pallet(Example)]
+pub mod pallet {
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use frame_system::ensure_root;
+
+	#[pallet::trait_]
+	pub trait Trait<I: Instance = DefaultInstance>: frame_system::Trait {
+		type Balance: Parameter + codec::HasCompact + From<u32> + Into<Weight> + Default
+			+ MaybeSerializeDeserialize;
+		#[pallet::const_]
+		type SomeConst: Get<Self::Balance>;
+		type Event: From<Event<Self, I>> + IsType<<Self as frame_system::Trait>::Event>;
+	}
+
+	#[pallet::module]
+	pub struct Module<T, I = DefaultInstance>(PhantomData<(T, I)>);
+
+	#[pallet::module_interface]
+	impl<T: Trait<I>, I: Instance> ModuleInterface<T::BlockNumber> for Module<T, I> {
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			<Dummy<T, I>>::put(T::Balance::from(10));
+			10
+		}
+
+		fn on_finalize(_n: T::BlockNumber) {
+			<Dummy<T, I>>::put(T::Balance::from(11));
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Trait<I>, I: Instance> Module<T, I> {
+		#[pallet::weight(<T::Balance as Into<Weight>>::into(new_value.clone()))]
+		fn set_dummy(origin: OriginFor<T>, #[pallet::compact] new_value: T::Balance) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+
+			<Dummy<T, I>>::put(&new_value);
+			Self::deposit_event(Event::Dummy(new_value));
+
+			Ok(().into())
+		}
+	}
+
+	#[pallet::error]
+	pub enum Error<T, I = DefaultInstance> {
+		/// Some wrong behavior
+		Wrong,
+	}
+
+	#[pallet::event]
+	#[pallet::generate(fn deposit_event)]
+	pub enum Event<T: Trait<I>, I: Instance = DefaultInstance> {
+		/// Dummy event, just here so there's a generic type that's used.
+		Dummy(T::Balance),
+	}
+
+	#[pallet::storage]
+	/// Some documentation
+	type Dummy<T: Trait<I>, I: Instance = DefaultInstance> = StorageValueType<_, T::Balance, OptionQuery>;
+
+	#[pallet::storage]
+	type Bar<T: Trait<I>, I: Instance = DefaultInstance> =
+		StorageMapType<_, Blake2_128Concat, T::AccountId, T::Balance, ValueQuery>;
+
+	#[pallet::storage]
+	type Foo<T: Trait<I>, I: Instance = DefaultInstance> =
+		StorageValueType<_, T::Balance, ValueQuery, OnFooEmpty<T, I>>;
+	#[pallet::type_value] pub fn OnFooEmpty<T: Trait<I>, I: Instance>() -> T::Balance { 3.into() }
```

Comment:
What is a `type_value`?

Reply:
the idea came from Gav https://github.com/paritytech/substrate/issues/5678#issuecomment-679934805 as an easy way to implement a struct which implements `Get`.

It is same as `parameter_types` but it allows to have some generic. another possibility is to extend parameter type to allow generic, and it should be directly usable here.
The macro check that generic are consistent (i.e. instance generic is used or not depending on trait) but anyway code will not compile if user mixed himself.

Reply:
Ahh it's for setting the default value, I see. Hmm.